Zack/mapnodes

Zhirschhorn opened the pull request:
Changes: 
-I removed the intermediate nodes from the selection screen
-I removed the dashed line
-I removed the paths that the bots are currently traveling on
-I added the custom icons to the buttons
Did not change:
-Did not add the A* path to the node from the bot. There needs to be a backend change as well because right now you can't get a path between two spots -- only send a bot to a spot
-when the bot is sent change the line to blue to update where the bot is (sending bots is broken)
-change the icons of the locations from the pins, not sure what we want them to look like before sending the bot

Zhirschhorn commented on src/containers/Map/MapScreen.tsx:
```diff
@@ -341,7 +335,7 @@ const formatEventBotsData = (
 			items.push({ ...obj.item, quantity: obj.quantity, bot: bot });
 			itemCount += obj.quantity;
-			itemsSold += obj.sales.numSold;
+			//itemsSold += obj.sales.numSold;
```
For some reason this line caused the code to break so I commented it out. Should add to bug report

Saagar3690 commented on src/containers/Map/MapScreen.tsx:
```diff
@@ -365,38 +359,36 @@ const formatEventBotsData = (
+		});
 	return { mapNodeArray: mapNodeMarkers, mapNodeHeaderInfo };
 };
 
```
Do we need this function formatMapPathsData if we don't have paths anymore? Or we keeping it because we still need it for ordering a bot?

Zhirschhorn replied:
keep it for ordering a bot

Saagar3690 reviewed (changes requested):
Looks good. Just some minor things, basically removing any unused things now just so that the code is cleaner. And regarding the line where it caused to crash, make sure on the backend you are migrating up both setup and sales because that should make it work.

Zhirschhorn commented on src/containers/Map/MapScreen.tsx:
```diff
@@ -39,9 +39,9 @@ const MapScreen = () => {
-	const [botPaths, setBotPaths] = useState<Location[][] | null>(null);
+
+	// Path between selected Bot and selected Location
+	const [paths, setPaths] = useState<Location[][] | null>(null);
```
We should leave it in because we will probably use it later. Its used to draw lines on the map, and we want to keep that functionality (just temporaritly not being used)

Zhirschhorn commented on src/containers/Map/MapView.tsx:
```diff
@@ -14,9 +14,13 @@ import React, { useEffect, useRef, useState } from 'react';
 import { MAP_MARKER_SIZE } from '../../constants';
 import { MarkerData, PropTypes } from './mapTypes';
 // import mapDest from '../../assets/mapDest.png';
+import hamburgerIcon from '../../assets/ICON_hamburger_menu.png';
```
nope

Zhirschhorn commented:
Ok so totally revamped this change. All the functionality should be working, but some visually things (like the popups) are not. I just wanted to get this merged in and take a break before I come back and tackle the visuals (because I hate CSS lol)

Zhirschhorn commented:
Changes:
-Added a type to the marker props (will either be 'bot', 'mapnode' (for unadded), 'N' (a number which it is on the route)
-Added a function addToRoute which is called whenever a node is either added or removed -- it recalculates the needed entries and rerenders
-Code to add the callout to the mapnodes when pressed
-code to display the nodes differently when part of route or not , or selected or not
-Added MapScreenProp botSelected so we know if there is a bot already selected or not (this is to integrate with thilans code)
-probably some more small stuff :)

**Screenshots**
![WhatsApp Image 2021-04-17 at 2 30 35 PM](https://user-images.githubusercontent.com/13991762/115127226-c6d33b00-9f89-11eb-85dd-6cf56c234e47.jpeg)
![WhatsApp Image 2021-04-17 at 2 30 34 PM (3)](https://user-images.githubusercontent.com/13991762/115127230-cc308580-9f89-11eb-944d-7ae01eedf296.jpeg)
![WhatsApp Image 2021-04-17 at 2 30 34 PM (2)](https://user-images.githubusercontent.com/13991762/115127232-d18dd000-9f89-11eb-9e09-9c7c31bbe380.jpeg)
![WhatsApp Image 2021-04-17 at 2 30 34 PM (1)](https://user-images.githubusercontent.com/13991762/115127233-d2befd00-9f89-11eb-9950-324404e557b7.jpeg)
![WhatsApp Image 2021-04-17 at 2 30 34 PM](https://user-images.githubusercontent.com/13991762/115127234-d3579380-9f89-11eb-84dd-b4cde113f503.jpeg)

Zhirschhorn commented:
As you can see there is more design to fix with colors, sizing, and especially the callout

thilan-tran commented on src/containers/Map/MapView.tsx:
```diff
@@ -11,12 +19,34 @@ import MapView, {
-import { MarkerData, PropTypes } from './mapTypes';
+//import { styles as formStyles } from '../Auth/FormStyles';
 // import mapDest from '../../assets/mapDest.png';
+//import { exp } from 'react-native-reanimated';
```
Quick reminder to get rid of commented imports / console logs

thilan-tran replied:
Quick reminder to get rid of commented imports / console logs

thilan-tran commented on src/containers/Map/MapScreen.tsx:
```diff
@@ -50,36 +54,95 @@ const MapScreen = () => {
+			await removeFromRoute(marker);
+		} else {
+			if (curRoute.length === 0 && selectedBotForOrder) {
+				let curPaths = paths ? paths : [];
```
Tiny nitpick, for this common ternary pattern that shows up a couple times, you can also use `let curPaths = paths || [];` (if you want 😄 )

In JS `||, &&` act as selection operators in addition to boolean ones.

thilan-tran commented on src/containers/Map/MapScreen.tsx:
```diff
@@ -213,6 +274,7 @@ const MapScreen = () => {
 	}
 
 	if (showMapNodes && selectedBotForOrder) {
+		let a: boolean = rerender;
```
What exactly is the `rerender` state being used for? This is the only instance it is used, as far as I can tell.

Zhirschhorn replied:
Basically when I add a route to the current route array the app won't rerender and display the changes. So it was kinda a hacky fix I added to change this state so the page would reload and the changes would display. Idk if you have a better idea

thilan-tran replied:
I bet that's exactly because of the use of array mutating methods I mentioned. If you change the array methods you use to `concat` etc., React should detect the array has changed and rerender automatically.

Zhirschhorn replied:
you were right. I changed it and took out the rerender. Thanks for letting me know!

thilan-tran commented on src/containers/Map/MapScreen.tsx:
```diff
@@ -50,36 +54,95 @@ const MapScreen = () => {
+	}
+
+	async function removeFromRoute(marker: MarkerData) {
+		console.log('adding marker ' + marker.name);
```
Should we print `'remove marker'` here?

Zhirschhorn replied:
ya

thilan-tran commented on src/containers/Map/MapScreen.tsx:
```diff
@@ -50,36 +54,95 @@ const MapScreen = () => {
+				);
+				curPaths.push(newPath);
+				setPaths(curPaths);
+			} else if (curRoute) {
```
You might want this if condition to be `curRoute.length`, all objects in JS are truthy, even empty arrays and objects.

thilan-tran commented on src/containers/Map/MapScreen.tsx:
```diff
@@ -50,36 +54,95 @@ const MapScreen = () => {
+					selectedBotForOrder.location,
+					marker.location
+				);
+				curPaths.push(newPath);
```
React nitpick, generally React suggests avoiding using in-place array modification methods like `push, splice` on state.
ie. you never want to mutate the state directly, even if you call `setState` on it after.

Here, `curPaths` refers to the same `paths` object, so if we then run `setPaths(curPaths)` after modifying the array in-place, there can be some undesired side effects

Instead they recommend to use modifications methods that create new arrays for example `concat` for inserts,
and `slice, filter` for removing elements like [here](https://stackoverflow.com/a/36326808).

thilan-tran reviewed (changes requested):
LGTM! Just a couple nitpicks that I had. (some JS and React things)

zhang-edward commented on src/containers/Map/MapView.tsx:
```diff
@@ -197,15 +234,15 @@ const MapComponent = ({
 						);
 					})} */}
-				{selected && centralMarker && (
+				{/*selected && centralMarker && (
```
Should delete these instead of commenting out?

katherinwang reviewed (changes requested):
Is there any way to make it so that the bot path goes under potential pinpoints/ user-set locations? In other words, is it possible to center the black circles (path crosses through midpoint) and for the gray pin icons bottom points to be bottom-aligned? 

Other than that, the bottom thingy won't be there right?

Zhirschhorn commented:
> Is there any way to make it so that the bot path goes under potential pinpoints/ user-set locations? In other words, is it possible to center the black circles (path crosses through midpoint) and for the gray pin icons bottom points to be bottom-aligned?
> 
> Other than that, the bottom thingy won't be there right?

bottom thingy wont be there and I'll try to get the black circles at the midpoint on my next pr. just trying to get this merged

thilan-tran reviewed (approved):
LGTM!
